plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonPageSource.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.hive.ion;
+
+import com.amazon.ion.IonBufferConfiguration;
+import com.amazon.ion.IonReader;
+import com.amazon.ion.IonType;
+import io.trino.hive.formats.ion.IonDecoder;
+import io.trino.spi.Page;
+import io.trino.spi.PageBuilder;
+import io.trino.spi.connector.ConnectorPageSource;
+
+import java.io.IOException;
+import java.util.OptionalLong;
+import java.util.function.LongSupplier;
+
+import static io.airlift.slice.SizeOf.instanceSize;
+
+public class IonPageSource
+        implements ConnectorPageSource
+{
+    private static final int INSTANCE_SIZE = instanceSize(IonPageSource.class);
+
+    private final IonReader ionReader;
+    private final LongSupplier counter;
+    private final PageBuilder pageBuilder;
+    private final IonDecoder decoder;
+    private long readTimeNanos;
+    private int completedPositions;
+    private boolean finished;
+
+    public IonPageSource(IonReader ionReader, LongSupplier counter, IonDecoder decoder, PageBuilder pageBuilder)
+    {
+        this.ionReader = ionReader;
+        this.counter = counter;
+        this.decoder = decoder;
+        this.pageBuilder = pageBuilder;
```

Comment:
With the change I suggested to IonDecoder, the page build can be constructed here with:
```
this.pageBuilder = new PageBuilder(decoder.getTypes());